src/org-folding-and-outline-provider.ts

```diff
@@ -0,0 +1,126 @@
+import {
+    FoldingRangeProvider,
+    DocumentSymbolProvider,
+    CancellationToken,
+    TextDocument,
+    FoldingRange,
+    SymbolInformation,
+    ProviderResult,
+    SymbolKind,
+    Range,
+    Position
+} from 'vscode';
+
+// ChunkType value is used as SymbolKind for outline
+enum ChunkType {
+    SECTION = SymbolKind.Constant,
+    BLOCK = SymbolKind.Number
+}
+type Chunk = { type: ChunkType, title: string, level: number, startLine: number };
+
+export class OrgFoldingAndOutlineProvider implements FoldingRangeProvider, DocumentSymbolProvider {
+
+    private documentStateRegistry: WeakMap<TextDocument, OrgFoldingAndOutlineDocumentState>;;
+
+    constructor() {
+        this.documentStateRegistry = new WeakMap()
+    }
+
+    provideFoldingRanges(document: TextDocument, token: CancellationToken): ProviderResult<FoldingRange[]> {
+        let state = this.getOrCreateDocumentState(document);
+        return state.getRanges(document);
+    }
+
+    provideDocumentSymbols(document: TextDocument, token: CancellationToken): ProviderResult<SymbolInformation[]> {
+        let state = this.getOrCreateDocumentState(document);
+        return state.getSymbols(document);
+    }
+
+    private getOrCreateDocumentState(document: TextDocument): OrgFoldingAndOutlineDocumentState {
+        let state = this.documentStateRegistry.get(document);
+        if (!state) {
+            state = new OrgFoldingAndOutlineDocumentState();
+            this.documentStateRegistry.set(document, state);
+        }
+        return state;
+    }
+}
+
+class OrgFoldingAndOutlineDocumentState {
+    private computedForDocumentVersion: number = null;
+    private ranges: FoldingRange[] = [];
+    private symbols: SymbolInformation[] = [];
+
+    getRanges(document: TextDocument): FoldingRange[] {
+        this.compute(document);
+        return this.ranges;
+    }
+
+    getSymbols(document: TextDocument): SymbolInformation[] {
+        this.compute(document);
+        return this.symbols;
+    }
+
+    private compute(document: TextDocument) {
+        if (document.version === this.computedForDocumentVersion) {
+            return;
+        }
+        this.computedForDocumentVersion = document.version;
+        this.ranges = [];
+        this.symbols = [];
+
+        const count = document.lineCount;
+        let stack: Chunk[] = [];
+        let inBlock = false;
+
+        for (let lineNumber = 0; lineNumber < count; lineNumber++) {
+            const element = document.lineAt(lineNumber);
+            const text = element.text;
+
+            if (inBlock) {
+                // we look for the end of the block
+                if (/^ *#\+END_/i.test(text)) {
```

Comment:
Hi @sipi, what do you think about moving such checks into methods like `isBlockEnd`, `isBlockStart`, `isHeader`? I think this will improve readability and make comments redundant, also they could be usable for future use.

Reply:
I totally agree. I'm use to say: if you have to comment inside the method body, create a  new method instead and comment it if needed. ;)

Reply:
Thanks for this review

Reply:
Done, and I reused (and rewrote) ``getStarPrefixCount`` (I'm not really happy with the ``get`` because this method ``compute``, so ``computeStarPrefixCount``would be more appropriate… But, it is not the topic of this PR)